sites/game-client/ui/main-ui.mjs

```diff
@@ -58,14 +71,72 @@ class MainUI {
    */
   setMainView(overlay) {}
 
-  // TODO: Stack visualization? Stack height limit?
   /**
    * Show notification in notification area.
-   * @param msg The message want to be shown.
+   * @param {String} msg The message want to be shown.
    * @param {Number} timeout The duraion of showing notification, the unit is millisecond.
    */
   showNotification(msg, timeout) {
     // TODO(lisasasasa)
+    // Every element of the list is {ele, next}, where `ele` is DOMElement
+    this._msgList ??= new LinkedList();
+
+    // An arrow function to remove the notification
+    const removeNotification = (listNode, clear) => {
+      const {ele, timer} = listNode.content;
+      if (clear) clearTimeout(timer);
+      if (this._msgList.delete(listNode)) this.notificationDom.removeChild(ele);
+    };
+
+    // Before inserting the new notification, check if list exceed the limit
+    if (this._msgList.length === NOTIFICATION_NUM_LIMIT) {
+      removeNotification(this._msgList.findNode(NOTIFICATION_NUM_LIMIT), true);
+    }
+
+    const ele = document.createElement('div');
+    const msgEle = document.createTextNode(msg);
+    ele.appendChild(msgEle);
+
+    this.notificationDom.appendChild(ele);
+    const content = {ele};
+    const listNode = this._msgList.insert(content);
+
+    const timer = setTimeout(() =>{
+      removeNotification.bind(this, listNode, false)();
+    }, timeout);
+    listNode.content.timer = timer;
+    console.log(this._msgList)
+  }
+
+  /**
+   * Show announcement in marquee.
+   * @param {String} msg The message want to be shown.
+   * @param {Number} timeout The duraion of showing notification, the unit is millisecond.
+   */
+  showAnnouncement(msg, timeout) {
+    // Every element of the list is {sting, next}
+    this._announceList ??= new LinkedList();
+    // TODO(marquee)
+    // An arrow function to remove the notification
+    const removeNotification = (listNode, clear) => {
+      const {msg, timer} = listNode.content;
+      if (clear) clearTimeout(timer);
+      if (this._announceList.delete(listNode)) {
+        this.announcementWord.textContent = this._announceList.findAllElement().join('  ');
+      }
+    };
+
+    if (this._announceList.length === NOTIFICATION_NUM_LIMIT) {
+      removeNotification(this._announceList.findNode(NOTIFICATION_NUM_LIMIT), true);
+    }
+
+    const content = {msg};
+    const listNode = this._announceList.insert(content);
+    this.announcementWord.textContent = this._announceList.findAllElement().join('  ');
```

Comment:
Nit, optional: Separate this findAllElement().join() part into a separate function so we'll repeat lesser code?